From 3061b1b45930c86796da113f152e716a4bbdb618 Mon Sep 17 00:00:00 2001
From: Benedek Kupper <kupper.benedek@gmail.com>
Date: Wed, 19 Feb 2025 12:32:11 +0100
Subject: [PATCH] bluetooth: dis: allow more dynamic serial number handling

---
 subsys/bluetooth/services/dis.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/subsys/bluetooth/services/dis.c b/subsys/bluetooth/services/dis.c
index 27676600b1..7f01d0c1c6 100644
--- a/subsys/bluetooth/services/dis.c
+++ b/subsys/bluetooth/services/dis.c
@@ -222,6 +222,13 @@ static ssize_t read_udi(struct bt_conn *conn, const struct bt_gatt_attr *attr, v
 }
 #endif
 
+#if defined(CONFIG_BT_DIS_SERIAL_NUMBER)
+extern char serial_number_str[];
+char *const serial_number_ptr =
+	(sizeof(BT_DIS_SERIAL_NUMBER_STR_REF) <= 1 ? serial_number_str
+						   : BT_DIS_SERIAL_NUMBER_STR_REF);
+#endif
+
 /* Device Information Service Declaration */
 BT_GATT_SERVICE_DEFINE(
 	dis_svc, BT_GATT_PRIMARY_SERVICE(BT_UUID_DIS),
@@ -237,7 +244,7 @@ BT_GATT_SERVICE_DEFINE(
 
 #if defined(CONFIG_BT_DIS_SERIAL_NUMBER)
 	BT_GATT_CHARACTERISTIC(BT_UUID_DIS_SERIAL_NUMBER, BT_GATT_CHRC_READ, BT_GATT_PERM_READ,
-			       read_str, NULL, BT_DIS_SERIAL_NUMBER_STR_REF),
+			       read_str, NULL, serial_number_ptr),
 #endif
 #if defined(CONFIG_BT_DIS_FW_REV)
 	BT_GATT_CHARACTERISTIC(BT_UUID_DIS_FIRMWARE_REVISION, BT_GATT_CHRC_READ, BT_GATT_PERM_READ,
-- 
2.44.0

