From 94ce17290aab6e4b17fa5d3fce19413d19d93e48 Mon Sep 17 00:00:00 2001
From: Benedek Kupper <kupper.benedek@gmail.com>
Date: Fri, 1 Aug 2025 23:32:59 +0200
Subject: [PATCH] drivers: udc: avoid infinite loop when closing OUT EPs

---
 drivers/usb/udc/udc_common.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/usb/udc/udc_common.c b/drivers/usb/udc/udc_common.c
index 3718290ce4..597baefe4f 100644
--- a/drivers/usb/udc/udc_common.c
+++ b/drivers/usb/udc/udc_common.c
@@ -123,6 +123,10 @@ struct net_buf *udc_buf_get_all(struct udc_ep_config *const ep_cfg)
 	LOG_DBG("ep 0x%02x dequeue %p", ep_cfg->addr, buf);
 	for (struct net_buf *n = buf; !k_fifo_is_empty(&ep_cfg->fifo); n = n->frags) {
 		n->frags = k_fifo_get(&ep_cfg->fifo, K_NO_WAIT);
+		if (n->frags == n) {
+			LOG_WRN("ep 0x%02x dequeue FIFO corrupt: %p", ep_cfg->addr, n);
+			n->frags = NULL;
+		}
 		LOG_DBG("|-> %p ", n->frags);
 		if (n->frags == NULL) {
 			break;
-- 
2.43.0

